osquery/tables/system/posix/smbios_utils.cpp

```diff
@@ -577,6 +577,34 @@ void genSMBIOSOEMStrings(const SMBStructHeader* hdr,
   }
 }
 
+void genSMBIOSProcessor(size_t index,
+                        const SMBStructHeader* hdr,
+                        uint8_t* address,
+                        uint8_t* textAddrs,
+                        size_t size,
+                        QueryData& results) {
+  const size_t maxOffset = 0x2e + 2;
+  if (hdr->type != kSMBIOSTypeProcessor || size < maxOffset || index < 4) {
```

Comment:
`index < 4`, this is not correct. On my Linux KVM, after adding a couple of additional logs:

```
I0429 15:21:18.001973 210842 smbios_tables.cpp:105] Reading SMBIOS from sysfs DMI node
I0429 15:21:18.001999 210842 smbios_utils.cpp:588] Header type: 0
I0429 15:21:18.002005 210842 smbios_utils.cpp:589] Size: 63
I0429 15:21:18.002008 210842 smbios_utils.cpp:590] Index: 0
I0429 15:21:18.002012 210842 smbios_utils.cpp:591] 
I0429 15:21:18.002013 210842 smbios_utils.cpp:588] Header type: 1
I0429 15:21:18.002017 210842 smbios_utils.cpp:589] Size: 75
I0429 15:21:18.002019 210842 smbios_utils.cpp:590] Index: 1
I0429 15:21:18.002022 210842 smbios_utils.cpp:591] 
I0429 15:21:18.002024 210842 smbios_utils.cpp:588] Header type: 3
I0429 15:21:18.002027 210842 smbios_utils.cpp:589] Size: 39
I0429 15:21:18.002030 210842 smbios_utils.cpp:590] Index: 2
I0429 15:21:18.002033 210842 smbios_utils.cpp:591] 
I0429 15:21:18.002035 210842 smbios_utils.cpp:588] Header type: 4 <-----
I0429 15:21:18.002038 210842 smbios_utils.cpp:589] Size: 65
I0429 15:21:18.002041 210842 smbios_utils.cpp:590] Index: 3
I0429 15:21:18.002044 210842 smbios_utils.cpp:591] 
I0429 15:21:18.002054 210842 smbios_utils.cpp:588] Header type: 16
I0429 15:21:18.002064 210842 smbios_utils.cpp:589] Size: 25
I0429 15:21:18.002068 210842 smbios_utils.cpp:590] Index: 4
I0429 15:21:18.002074 210842 smbios_utils.cpp:591] 
I0429 15:21:18.002082 210842 smbios_utils.cpp:588] Header type: 17
I0429 15:21:18.002085 210842 smbios_utils.cpp:589] Size: 53
I0429 15:21:18.002089 210842 smbios_utils.cpp:590] Index: 5
I0429 15:21:18.002091 210842 smbios_utils.cpp:591] 
I0429 15:21:18.002094 210842 smbios_utils.cpp:588] Header type: 17
I0429 15:21:18.002096 210842 smbios_utils.cpp:589] Size: 53
I0429 15:21:18.002099 210842 smbios_utils.cpp:590] Index: 6
I0429 15:21:18.002102 210842 smbios_utils.cpp:591] 
I0429 15:21:18.002104 210842 smbios_utils.cpp:588] Header type: 17
I0429 15:21:18.002110 210842 smbios_utils.cpp:589] Size: 53
I0429 15:21:18.002112 210842 smbios_utils.cpp:590] Index: 7
I0429 15:21:18.002115 210842 smbios_utils.cpp:591] 
I0429 15:21:18.002118 210842 smbios_utils.cpp:588] Header type: 19
I0429 15:21:18.002120 210842 smbios_utils.cpp:589] Size: 33
I0429 15:21:18.002125 210842 smbios_utils.cpp:590] Index: 8
I0429 15:21:18.002126 210842 smbios_utils.cpp:591] 
I0429 15:21:18.002130 210842 smbios_utils.cpp:588] Header type: 19
I0429 15:21:18.002131 210842 smbios_utils.cpp:589] Size: 33
I0429 15:21:18.002133 210842 smbios_utils.cpp:590] Index: 9
I0429 15:21:18.002136 210842 smbios_utils.cpp:591] 
I0429 15:21:18.002138 210842 smbios_utils.cpp:588] Header type: 32
I0429 15:21:18.002140 210842 smbios_utils.cpp:589] Size: 13
I0429 15:21:18.002143 210842 smbios_utils.cpp:590] Index: 10
I0429 15:21:18.002146 210842 smbios_utils.cpp:591] 
I0429 15:21:18.002147 210842 smbios_utils.cpp:588] Header type: 127
I0429 15:21:18.002151 210842 smbios_utils.cpp:589] Size: 4
I0429 15:21:18.002152 210842 smbios_utils.cpp:590] Index: 11
I0429 15:21:18.002154 210842 smbios_utils.cpp:591] 
```
As you can see the processor `Header type: 4` appears at `Index: 3`, in fact the table doesn't return any result.

It seems to me you're doing this check just to later be able to give a number to the CPU (which is also not guaranteed to start from 0).

I would remove this logic and instead check in the caller of this function if there is a row that has been generated, after it has been called in the lambda. If so, then set the `device_id` with the value of `CPU` +  a counter you increase inside the lambda.

Reply:
Thanks for the comment. I've applied the change you requested.

Reply:
@Smjert, Can you take a look at my latest fix please?